File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> result; // To store the final zigzag order traversal
        if (!root) return result; // If the tree is empty, return an empty result

        queue<TreeNode*> q; // Queue to hold nodes for BFS
        q.push(root); // Start with the root node
        bool leftToRight = true; // Flag to track direction of traversal

        while (!q.empty()) {
            int levelSize = q.size(); // Number of nodes at the current level
            vector<int> currentLevel(levelSize); // To store values of the current level

            for (int i = 0; i < levelSize; ++i) {
                TreeNode* currentNode = q.front(); // Get the front node
                q.pop(); // Remove it from the queue
                
                // Determine position based on direction
                if (leftToRight) {
                    currentLevel[i] = currentNode->val; // Left to right
                } else {
                    currentLevel[levelSize - 1 - i] = currentNode->val; // Right to left
                }

                // Add children to the queue for the next level
                if (currentNode->left) q.push(currentNode->left);
                if (currentNode->right) q.push(currentNode->right);
            }

            result.push_back(currentLevel); // Add current level to the result
            leftToRight = !leftToRight; // Toggle direction for next level
        }

        return result; // Return the final result
    }
};

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>(); // To store the final zigzag order traversal
        if (root == null) return result; // If the tree is empty, return an empty result

        Queue<TreeNode> queue = new LinkedList<>(); // Queue to hold nodes for BFS
        queue.offer(root); // Start with the root node
        boolean leftToRight = true; // Flag to track direction of traversal

        while (!queue.isEmpty()) {
            int levelSize = queue.size(); // Number of nodes at the current level
            List<Integer> currentLevel = new ArrayList<>(levelSize); // To store values of the current level

            for (int i = 0; i < levelSize; i++) {
                TreeNode currentNode = queue.poll(); // Get the front node
                
                // Determine position based on direction
                if (leftToRight) {
                    currentLevel.add(currentNode.val); // Left to right
                } else {
                    currentLevel.add(0, currentNode.val); // Right to left (add at the beginning)
                }

                // Add children to the queue for the next level
                if (currentNode.left != null) queue.offer(currentNode.left);
                if (currentNode.right != null) queue.offer(currentNode.right);
            }

            result.add(currentLevel); // Add current level to the result
            leftToRight = !leftToRight; // Toggle direction for next level
        }

        return result; // Return the final result
    }
}